### PR TITLE
Deprecate SXNET, some old NETSCAPE data types

### LINKED ISSUE #1177: remove NETSCAPE_CERT_SEQUENCE

### PR DESCRIPTION
**Pull request by richsalz**
Support for the SXNET and (Thawte Strong Extranet) certificate extension,
the NETSCAPE_CERT_SEQUENCE, NETSCAPE_SPKAC, and NETSCAPE_SPKI datatypes
have been deprecrated.

Fixes #1177, #10668

Deprecate things now so they can be removed when ready.

### CODE REVIEW

**Comment by richsalz** [comment]
Any clues appreciated.  You'd think I would know how to do deprecations by now.

**Comment by richsalz** [comment]
CI's pass, this is ready for review. It deprecates things that the project will want to remove in the future. From CHANGES:
```
 * Support for the SXNET (Thawte Strong Extranet) certificate extension,
   the NETSCAPE_CERT_SEQUENCE, NETSCAPE_SPKAC, and NETSCAPE_SPKI datatypes
   have been deprecrated.
```

File: CHANGES.md [hunk 1]
```diff
@@ -38,6 +39,12 @@ breaking changes, and mappings for the large list of deprecated functions.
 
    *Rich Salz*
 
+ * Support for the SXNET (Thawte Strong Extranet) certificate extension,
+   the NETSCAPE_CERT_SEQUENCE, NETSCAPE_SPKAC, and NETSCAPE_SPKI datatypes
+   have been deprecrated.
```
Comment (FdaSilvaYY) [inline CHANGES.md:44]: typo: deprecrated

Comment (richsalz) [reply CHANGES.md]: typo fixed; fixup commit pushed. thanks!

**Review by t8m — CHANGES_REQUESTED** [review]

File: apps/ca.c [hunk 2]
```diff
@@ -450,8 +454,10 @@ int ca_main(int argc, char **argv)
             req = 1;
             break;
         case OPT_SPKAC:
+#ifndef OPENSSL_NO_DEPRECATED_3_0
```
[the remark this answers is not in the document]
Comment (richsalz) [reply apps/ca.c:457]: No, `OPT_SPKAC` still shows in the enum list, so we need a `case` statement to handle it.  It's similar to how `OPT_ENGINE` is handled, where the enum is present, but the body ends up being no-ops for the `no-engine` config.

I think doing things this way minimizes the ifdef mazes and makes it clear that the code does nothing.  YMMV.

File: crypto/pem/pem_all.c [hunk 3]
```diff
@@ -41,8 +41,6 @@ IMPLEMENT_PEM_rw(X509_CRL, X509_CRL, PEM_STRING_X509_CRL, X509_CRL)
 IMPLEMENT_PEM_rw(X509_PUBKEY, X509_PUBKEY, PEM_STRING_PUBLIC, X509_PUBKEY)
 IMPLEMENT_PEM_rw(PKCS7, PKCS7, PEM_STRING_PKCS7, PKCS7)
 
-IMPLEMENT_PEM_rw(NETSCAPE_CERT_SEQUENCE, NETSCAPE_CERT_SEQUENCE,
-                 PEM_STRING_X509, NETSCAPE_CERT_SEQUENCE)
```
[the remark this answers is not in the document]
Comment (richsalz) [reply crypto/pem/pem_all.c:45]: `PEM_STRING_x509` is used in many places. Did you mean something else, like `pkey_get_rsa` etc?

File: crypto/asn1/nsseq.c [hunk 4]
```diff
@@ -7,6 +7,7 @@
  * https://www.openssl.org/source/license.html
  */
 
+#include "internal/deprecated.h"
```
[the remark this answers is not in the document]
Comment (richsalz) [reply crypto/asn1/nsseq.c:10]: Can you explain why the define is better than the header file?  And for bonus credit, when each should be used?
```
; g grep internal/deprecated.h | wc
    329     668   20577
; g grep OPENSSL_SUPPRESS_DEPRECATED | wc
     85     189    5061
```

File: doc/man3/X509V3_get_d2i.pod [hunk 5]
```diff
@@ -12,6 +12,10 @@ X509_REVOKED_add1_ext_i2d - X509 extension decode and encode functions
 
  #include <openssl/x509v3.h>
 
+Deprecated since OpenSSL 3.0, can be hidden entirely by defining
+B<OPENSSL_API_COMPAT> with a suitable version value, see
+L<openssl_user_macros(7)>:
+
```
[the remark this answers is not in the document]
Comment (richsalz) [reply doc/man3/X509V3_get_d2i.pod:18]: I have no idea why I did that.  Zombies.  Reverted, updated fixup commit pushed.

File: crypto/pem/pem_all.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (t8m) [reply crypto/pem/pem_all.c:45]: Just a confusion on my side when skimming through the PR.

File: apps/ca.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (t8m) [reply apps/ca.c:457]: OK

**Comment by richsalz** [comment]
@t8m: Should I create `crypto/asn1/asn1_depr` which has `t_spki` `x_spki` and `nsseq` ?

**Comment by t8m** [comment]
> @t8m: Should I create `crypto/asn1/asn1_depr` which has `t_spki` `x_spki` and `nsseq` ?

I do not think it is necessary.

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
OTC: We won't be doing this for 3.0. Setting the Post 3.0 milestone.

**Comment by richsalz** [comment]
As this has been moved to post-3.0, please remove the "approved" label; thanks.

**Pull request closed without merging**